### DISCUSSION TITLE
Is there a way to ignore column comments during `revision --autogenerate`

### DISCUSSION THREAD

**Discussion by zlisde** [sqlalchemy/alembic#892] [Usage Questions] [answered]
We are currently using alembic for schema change management, and also another tool to apply comments on columns for data documentation purpose.  However when we run `revision --autogenerate`,  alembic generate scripts removed all the comments not defined in the table models,  I am wondering if there is a way to ignore column comments during alembic change capture.

**Comment by zzzeek** [comment, marked as answer]
several, the most direct is to apply [column_reflect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=column_reflect#sqlalchemy.events.DDLEvents.column_reflect) to the Table class in env.py and set "comment" to blank:

```py
from sqlalchemy import event
from sqlalchemy import Table

@event.listens_for(Table, 'column_reflect')
def receive_column_reflect(inspector, table, column_info):
    column_info['comment'] = None
```

the compare logic should skip entirely if the "comment" is reflected as None so the above should disable all comparison.

**Reply by will-misslin** [reply]
This is great, thank you! Is there a similar way to intercept table comments (I couldn't find anything obvious after a bit of flailing) or would you recommend a different approach?

**Reply by zzzeek** [reply]
for both of these things you might be able to use the [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) hook, that will send in all the Table / Column / etc objects where you could also set ".comment = None".   This technique wont work for everything but for that it probably will.

**Reply by will-misslin** [reply]
Thank you for the response! It's amazing to have support for such a popular library directly from a creator!

Sorry, I am not quite understanding how to implement the include_object hook since my understanding is that it is essentially a boolean flag to either include or exclude a whole object. The documentation seems to indicate we can include/exclude only a whole object (one of `"table", "column", "index", "unique_constraint", or "foreign_key_constraint"`), not just exclude a specific object's type of attribute.

Can you help me understand? Also, I am happy to open a new discussion if needed! Thank you! 🙂

**Reply by zzzeek** [reply]
the idea is to repurpose the function as a way to intercept Column objects before they are compared:

```py
def include_object(object, name, type_, reflected, compare_to):
    # zero-out the "comment" field
    if type_ in ("column", "table"):
        object.comment = None

    # but continue to include the object in the autogenerate comparison
    return True

context.configure(
    # ...
    include_object = include_object
)
```

**Reply by zlisde** [reply]
I actually tried this,  it still detects comments in both column and table.  We are use snowflake,  so wondering if the dialect of snowflake behaves differently.

**Reply by zzzeek** [reply]
it shouldn't, the "include_object" hook is run first, then the "compare" is run after that and that is all within alembic.  if either .comment attribute is None on the reflected Column or model Column, it should skip the comparison.   same for table.

**Reply by zzzeek** [reply]
it looks like both comments have to be identical though, i.e. None for both model and reflected objet, or some other string

**Reply by will-misslin** [reply]
Is there a way for us to access the reflected object's metadata to set the model's comment to match the reflected object's comments? The `include_object` approach seems to be our best hope.

I have a hacky `post_write_hook` that uses regex to remove drop and create table comment directives, but it is vulnerable to breakage based on the contents of the table comments.

**Reply by CaselIT** [reply]
well `include_object` is usually defined inside the `env.py` file and it has access to the metadata https://github.com/sqlalchemy/alembic/blob/2c9656979ab8c4a670f0fdd019869a5ea20aa22e/alembic/templates/generic/env.py#L20

**Reply by zzzeek** [reply]
include_object is probably the only hook where you get the current and the compared object at the same time.  that said if you set all comments to None everywhere it should not come up with any comments being changed.

**Reply by will-misslin** [reply]
When I output the `target_metadata.tables` `comment` attribute for each table in the schema, all the results come back as `None`. To be extra sure, I set the object's comment to match the target_metadata value. This is the version of `include_object` I am using to debug:

```python
    def include_object(object, name, type_, reflected, compare_to):
        if type_ == "table":
			# print for debugging
            target_comment = target_metadata.tables[f"{SF_SCHEMA}.{object.name}"].comment
            print(f"object: {object.comment}")
            print(f'target: {target_comment}')

			# equate both versions to avoid triggering any comment ddl
            object.comment = target_comment
        # but continue to include the object in the autogenerate comparison
        return True
```


The output:
```
(.venv) > $ poetry run alembic --name data_analytics revision --autogenerate                       1s 

INFO  [alembic.runtime.migration] Context impl SnowflakeImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
object: None
target: None
...
```

However, the autogenerated revision still contains the table comment DDL. I also noticed that the `target_metadata.tables` object is an `immutable dict`, which sounds a bit ominous. Is there another attribute I should be using to inspect and set comments?

Thanks!

**Reply by zzzeek** [reply]
can you be specific?   it is illustrating "op.alter_table()" operations that change only the comment ?  can you look at the "reflected" boolean and "compare_to" to more accurately inspect/set the comments to be equivalent?  you probably want to set "comment" on both "object" and "compare_to" at the same time.

the "target_metadata" collection is the MetaData() object that you are passing to the environment, and the "tables" collection that's exposed is immutable because MetaData does not support table removal and table addition must go through the documented APIs.

**Reply by will-misslin** [reply]
Oh wow. I don't know how I missed it, but `compare_to` was absolutely what I was looking for!. Thanks for talking me through this! Here is the final version of `include_object` that we will use to remove both column and table level comments. With this, we can avoid also needing to use the `column_reflect` from the [original answer](https://github.com/sqlalchemy/alembic/discussions/892#discussioncomment-1235406).


```python
    def include_object(object, name, type_, reflected, compare_to):
        # capture tables and objects, excluding new model additions (nothing to
        # compare to)
        if type_ in ("table", "column") and compare_to is not None:
            # equate both versions to avoid triggering any comment ddl
            object.comment = compare_to.comment = None
        # but continue to include the object in the autogenerate comparison
        return True
```
